File: csharp/uwapi-cs/prototypes.cs
using System;
using System.Runtime.InteropServices;
using System.Collections.Generic;
using System.Text.Json;

namespace Unnatural
{
    public class ProtoResource
    {
        public uint id;
        public string name;
    }

    public class ProtoRecipe
    {
        public uint id;
        public string name;
        public Dictionary<uint, uint> inputs;
        public Dictionary<uint, uint> outputs;
        public uint duration;
        public uint placeOver;
    }

    public class ProtoConstruction
    {
        public uint id;
        public string name;
        public Dictionary<uint, uint> inputs;
        public uint output;
    }

    public class ProtoUnit
    {
        public uint id;
        public string name;
        public List<uint> recipes;
        public bool logistics;
        public bool assembler;
        public bool vital;
        public uint maxLife;
        public uint armorType;

        // weapon
        public uint damageType;
        public float dps;
        public float fireRange;
        public float rateOfFire;

        // mobile only
        public Dictionary<uint, float> speeds;
        public bool cargo;

        // building only
        public float buildingRadius;
    }

    public class HitChancesTable
    {
        public List<string> armorNames;
        public List<string> damageNames;
        public List<List<float>> hitChancesTable;
    }

    public class TerrainTypesTable
    {
        public List<string> terrainNames;
    }

    internal class Definitions
    {
        public HitChancesTable hitChancesTable;
        public TerrainTypesTable terrainTypesTable;
    }

    internal class ProtoGeneric
    {
        public uint type;
        public string name;
        public string json;
    }

    public static class Prototypes
    {
        public static IReadOnlyList<uint> All()
        {
            return all;
        }

        public static uint Type(uint id)
        {
            ProtoGeneric tmp;
            return types.TryGetValue(id, out tmp) ? tmp.type : 0;
        }

        public static string Name(uint id)
        {
            ProtoGeneric tmp;
            return types.TryGetValue(id, out tmp) ? tmp.name : "";
        }

        public static string Json(uint id)
        {
            ProtoGeneric tmp;
            return types.TryGetValue(id, out tmp) ? tmp.json : "";
        }

        public static ProtoResource Resource(uint id)
        {
            ProtoResource tmp;
            return resources.TryGetValue(id, out tmp) ? tmp : null;
        }

        public static ProtoRecipe Recipe(uint id)
        {
            ProtoRecipe tmp;
            return recipes.TryGetValue(id, out tmp) ? tmp : null;
        }

        public static ProtoConstruction Construction(uint id)
        {
            ProtoConstruction tmp;
            return constructions.TryGetValue(id, out tmp) ? tmp : null;
        }

        public static ProtoUnit Unit(uint id)
        {
            ProtoUnit tmp;
            return units.TryGetValue(id, out tmp) ? tmp : null;
        }

        public static HitChancesTable HitChancesTable()
        {
            return hitChancesTable;
        }

        public static TerrainTypesTable TerrainTypesTable()
        {
            return terrainTypesTable;
        }

        static readonly List<uint> all = new List<uint>();
        static readonly Dictionary<uint, ProtoGeneric> types = new Dictionary<uint, ProtoGeneric>();
        static readonly Dictionary<uint, ProtoResource> resources = new Dictionary<uint, ProtoResource>();
        static readonly Dictionary<uint, ProtoRecipe> recipes = new Dictionary<uint, ProtoRecipe>();
        static readonly Dictionary<uint, ProtoConstruction> constructions = new Dictionary<uint, ProtoConstruction>();
        static readonly Dictionary<uint, ProtoUnit> units = new Dictionary<uint, ProtoUnit>();
        static HitChancesTable hitChancesTable;
        static TerrainTypesTable terrainTypesTable;

        static uint[] AllIds()
        {
            Interop.UwIds ids = new Interop.UwIds();
            Interop.uwAllPrototypes(ref ids);
            return InteropHelpers.Ids(ids);
        }

        static void LoadPrototypes()
        {
            Console.WriteLine("loading prototypes");

            all.Clear();
            types.Clear();
            resources.Clear();
            recipes.Clear();
            constructions.Clear();
            units.Clear();

            var options = new JsonSerializerOptions
            {
                IncludeFields = true,
            };

            foreach (uint id in AllIds())
            {
                ProtoGeneric pg = new ProtoGeneric();
                pg.type = Interop.uwPrototypeType(id);
                pg.json = Marshal.PtrToStringAnsi(Interop.uwPrototypeJson(id));
                switch (pg.type)
                {
                    case 1:
                        resources[id] = JsonSerializer.Deserialize<ProtoResource>(pg.json, options);
                        pg.name = resources[id].name;
                        break;
                    case 2:
                        recipes[id] = JsonSerializer.Deserialize<ProtoRecipe>(pg.json, options);
                        pg.name = recipes[id].name;
                        break;
                    case 3:
                        constructions[id] = JsonSerializer.Deserialize<ProtoConstruction>(pg.json, options);
                        pg.name = constructions[id].name;
                        break;
                    case 4:
                        units[id] = JsonSerializer.Deserialize<ProtoUnit>(pg.json, options);
                        pg.name = units[id].name;
                        break;
                }
                types[id] = pg;
                all.Add(id);
            }

            Console.WriteLine("prototypes loaded");
        }

        static void LoadDefinitions()
        {
            Console.WriteLine("loading definitions");

            var options = new JsonSerializerOptions
            {
                IncludeFields = true,
            };

            string json = Marshal.PtrToStringAnsi(Interop.uwDefinitionsJson());
            Definitions defs = JsonSerializer.Deserialize<Definitions>(json, options);
            hitChancesTable = defs.hitChancesTable;
            terrainTypesTable = defs.terrainTypesTable;

            Console.WriteLine("definitions loaded");
        }

        static void Preparing(object sender, EventArgs e)
        {
            LoadPrototypes();
            LoadDefinitions();
        }

        static Prototypes()
        {
            Game.Preparing += Preparing;
        }
    }
}

File: python/uw/uw/prototypes.py
import json

from typing import Any
from typing import Optional

from .helpers import MapState
from .helpers import Prototype
from .helpers import _c_str


class ProtoGeneric:
    def __init__(self, _type: Prototype, name: str, _json: str):
        self.type = _type
        self.name = name
        self.json = _json


class Prototypes:
    def __init__(self, api, ffi, game):
        self._api = api,
        self._ffi = ffi
        self.game = game
        self.game.add_map_state_callback(self._map_state_changed)

        self._hit_chances_table: dict[str, Any]
        self._terrain_types_table: dict[str, Any]

        self._all: list[int] = []
        self._types: dict[int, ProtoGeneric] = {}
        self._resources: dict[str, Any] = {}
        self._recipes: dict[str, Any] = {}
        self._constructions: dict[str, Any] = {}
        self._units: dict[str, Any] = {}

    def all(self) -> list[int]:
        return self._all

    def type(self, _id: int) -> Prototype:
        return self._types[_id].type if _id in self._types else Prototype.NONE

    def name(self, _id: int) -> str:
        return self._types[_id].name if _id in self._types else ""

    def json(self, _id: int) -> str:
        return self._types[_id].json if _id in self._types else ""

    def resource(self, _id) -> Optional[dict]:
        return self._resources.get(_id)

    def recipes(self, _id) -> Optional[dict]:
        return self._recipes.get(_id)

    def construction(self, _id) -> Optional[dict]:
        return self._constructions.get(_id)

    def unit(self, _id) -> Optional[dict]:
        return self._units.get(_id)

    def hit_chances_table(self):
        return self._hit_chances_table

    def terrain_types_table(self):
        return self._terrain_types_table

    def all_ids(self) -> list[int]:
        ids = self._ffi.new("struct UwIds *")
        self._api.uwAllPrototypes(ids)
        return self._ffi.unpack(ids)

    def _load_prototypes(self):
        print("loading prototypes")

        self._all = []
        self._types = {}
        self._resources = {}
        self._recipes = {}
        self._constructions = {}
        self._units = {}

        for i in self.all_ids():
            _type = Prototype(self._api.uwPrototypeType(id))
            js = json.loads(_c_str(self._api.uwPrototypeJson(i)))
            if _type == Prototype.Resource:
                self._resources[id] = js
            elif _type == Prototype.Recipe:
                self._recipes[id] = js
            elif _type == Prototype.Construction:
                self._constructions[id] = js
            elif type == Prototype.Unit:
                self._units[id] = js

            self._types[id] = ProtoGeneric(_type, js["name"], js)
            self._all.append(id)

        print("prototypes loaded")

    def _load_definitions(self):
        print("loading definitions")

        defs = json.loads(_c_str(self._api.uwDefinitionsJson()))
        self._hit_chances_table = defs["hitChancesTable"]
        self._terrain_types_table = defs["terrainTypesTable"]

        print("definitions loaded")

    def _map_state_changed(self, state: MapState):
        if state == MapState.Loaded:
            self._load_prototypes()
            self._load_definitions()
